[MRG] FEA: Stacking estimator for classification and regression (#11047)
scikit-learn/scikit-learn: doc/modules/ensemble.rst

the VotingClassifier (with ``voting='hard'``) would classify the sample
as "class 1" based on the majority class label.

> In the cases of a tie, the :class:`VotingClassifier` will select the class based

Reviewer: This change should not be necessary now that we define `default_role`

.. topic:: Examples:

  * :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor.py`

.. _stacking:

> Stacked generalization

Reviewer 2: This is a subsection of the `VotingRegressor` part (1.11.6.1). Shouldn't this be in its own section (1.11.7) ?

Also nit: I think `Stacking` is more common and might be easier for the eye to catch

----------------------

Stacked generalization is a method for combining estimators to reduce their
biases [W1992]. More precisely, the predictions of each individual estimator

> are stacked together and used by another estimator to compute the final

Reviewer 2: I think we can make the "stacking" procedure a bit clearer with e.g.

```suggestion
are stacked together and used as input to a final estimator to compute the final
```

Before I looked at the code I thought the final estimator was learning some weights or something but it's more trivial than that.

prediction. This final estimator is trained through cross-validation.

The :class:`StackingClassifier` and :class:`StackingRegressor` provide such
strategies which can be applied to classification and regression problems.

The ``estimators`` parameter corresponds to the list of the estimators which
are stacked together. It should be given as a list of name and estimator::

  >>> from sklearn.linear_model import RidgeCV, LassoCV
  >>> from sklearn.svm import SVR
  >>> estimators = [('ridge', RidgeCV()),
  ...               ('lasso', LassoCV(random_state=42)),
  ...               ('svr', SVR(C=1, gamma=1e-6, kernel='rbf'))]

> The ``final_estimator`` will combine the predictions of the ``estimators``. It

Reviewer 2: *combine* suggests some aggregation procedure like weighting, I find it misleading. I would suggest

> The final estimator will use the predictions of the estimators as input.

needs to be a classifier or a regressor when using :class:`StackingClassifier`
or :class:`StackingRegressor`, respectively::

  >>> from sklearn.ensemble import GradientBoostingRegressor
  >>> from sklearn.ensemble import StackingRegressor
  >>> reg = StackingRegressor(
  ...     estimators=estimators,
  ...     final_estimator=GradientBoostingRegressor(random_state=42))

To train the ``estimators`` and ``final_estimator``, the ``fit`` method needs
to be called on the training data::

  >>> from sklearn.datasets import load_boston
  >>> X, y = load_boston(return_X_y=True)
  >>> from sklearn.model_selection import train_test_split
  >>> X_train, X_test, y_train, y_test = train_test_split(X, y,
  ...                                                     random_state=42)

>   >>> reg.fit(X_train, y_train)  # doctest: +ELLIPSIS

Reviewer 2: we can remove ELLIPSIS now \o/

  StackingRegressor(...)

During training, the ``estimators`` are fitted on the whole training data
``X_train``. They will be used when calling ``predict``, ``predict``, or
``predict_proba``. To generalize and avoid over-fitting, the
``final_estimator`` is trained on out-samples using internally
:func:`sklearn.model_selection.cross_val_predict`.

Note that the output of the ``estimators`` is controlled by the parameter

> ``method_estimators``. It corresponds to the method called by each

Reviewer 2: That's called `predict_method` now

estimator. This parameter is either a list of string of the methods name or

> ``'auto'`` which will automatically called the method depending of the

Reviewer 2: ```suggestion
``'auto'`` which will automatically call the method depending on the
```

availability and a pre-determined order of preference.

It is possible to get the final predictions using the ``predict`` or
``predict_proba`` depending if you are using :class:`StackingClassifier` or
:class:`StackingRegressor`::

>    >>> y_pred = reg.predict(X_test)

Reviewer 2: That example does not really illustrate the text above. May I suggest

> The fitted stacking estimators have a `predict` method, and classifier also have a `predict_proba` method. They can be used as any other estimator, e.g.::`

Or something like that (also, use single ticks)

   >>> from sklearn.metrics import r2_score
   >>> print('R2 score: {:.2f}'.format(r2_score(y_test, y_pred)))
   R2 score: 0.81

Note that it is also possible to get the output of the stacked outputs of the
``estimators`` using the ``transform`` method::

  >>> reg.transform(X_test[:5])  # doctest: +ELLIPSIS
  array([[28.78..., 28.43...  , 22.62...],
         [35.96..., 32.58..., 23.68...],
         [14.97..., 14.05..., 16.45...],
         [25.19..., 25.54..., 22.92...],
         [18.93..., 19.26..., 17.03... ]])

.. note::

>    Multiple layers stacking can be achieved by assigning ``final_estimator`` to

Reviewer 2: ```suggestion
   Multiple stacking layers can be achieved by assigning ``final_estimator`` to
```

   a :class:`StackingClassifier` or :class:`StackingRegressor`.

.. topic:: References

   .. [W1992] Wolpert, David H. "Stacked generalization." Neural networks 5.2
      (1992): 241-259.

>    >>> eclf = VotingClassifier(

Reviewer 2: remove ?

.. topic:: Examples:

  * :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor.py`

.. _stacking:

Stacked generalization
----------------------

Stacked generalization is a method for combining estimators to reduce their
biases [W1992]. More precisely, the predictions of each individual estimator
are stacked together and used by another estimator to compute the final

> prediction. This final estimator is trained through cross-validation.

Author: ```
This final estimator is trained on the predictions obtained from rolling left-out samples using  :func:`cross_val_predict`.
```

  >>> from sklearn.ensemble import GradientBoostingRegressor
  >>> from sklearn.ensemble import StackingRegressor
  >>> reg = StackingRegressor(
  ...     estimators=estimators,
  ...     final_estimator=GradientBoostingRegressor(random_state=42))

To train the ``estimators`` and ``final_estimator``, the ``fit`` method needs
to be called on the training data::

  >>> from sklearn.datasets import load_boston
  >>> X, y = load_boston(return_X_y=True)
  >>> from sklearn.model_selection import train_test_split
  >>> X_train, X_test, y_train, y_test = train_test_split(X, y,
  ...                                                     random_state=42)
  >>> reg.fit(X_train, y_train)  # doctest: +ELLIPSIS
  StackingRegressor(...)

During training, the ``estimators`` are fitted on the whole training data
``X_train``. They will be used when calling ``predict``, ``predict``, or
``predict_proba``. To generalize and avoid over-fitting, the
``final_estimator`` is trained on out-samples using internally
:func:`sklearn.model_selection.cross_val_predict`.

Note that the output of the ``estimators`` is controlled by the parameter
``method_estimators``. It corresponds to the method called by each

> estimator. This parameter is either a list of string of the methods name or

Reviewer: Or rephrase

.. topic:: Examples:

  * :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor.py`

.. _stacking:

Stacked generalization
======================

Stacked generalization is a method for combining estimators to reduce their
biases [W1992]. More precisely, the predictions of each individual estimator
are stacked together and used as input to a final estimator to compute the
final prediction. This final estimator is trained through cross-validation.

The :class:`StackingClassifier` and :class:`StackingRegressor` provide such
strategies which can be applied to classification and regression problems.

The `estimators` parameter corresponds to the list of the estimators which

> are stacked together. It should be given as a list of name and estimator::

Reviewer 3: ```suggestion
are stacked together. It should be given as a list of names and estimators::
```

  >>> reg = StackingRegressor(
  ...     estimators=estimators,
  ...     final_estimator=GradientBoostingRegressor(random_state=42))

To train the `estimators` and `final_estimator`, the `fit` method needs
to be called on the training data::

  >>> from sklearn.datasets import load_boston
  >>> X, y = load_boston(return_X_y=True)
  >>> from sklearn.model_selection import train_test_split
  >>> X_train, X_test, y_train, y_test = train_test_split(X, y,
  ...                                                     random_state=42)
  >>> reg.fit(X_train, y_train)
  StackingRegressor(...)

During training, the `estimators` are fitted on the whole training data
`X_train`. They will be used when calling `predict`, `predict`, or
`predict_proba`. To generalize and avoid over-fitting, the
`final_estimator` is trained on out-samples using internally
:func:`sklearn.model_selection.cross_val_predict`.

Note that the output of the `estimators` is controlled by the parameter
`predict_method`. It corresponds to the method called by each
estimator. This parameter is either a list of string of the methods name or
`'auto'` which will automatically call the method depending on the

> availability and a pre-determined order of preference.

Reviewer 3: Should we specify the order of preference?

Optionally, weights can be provided for the individual classifiers::

   >>> eclf = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)], voting='soft', weights=[2,5,1])

.. _stacking:

Stacked generalization
----------------------

Stacked generalization is a method for combining estimators to reduce their
biases [W1992]. More precisely, the predictions of each individual estimator
are stacked together and used by another estimator to compute the final
prediction. This final estimator is trained through cross-validation.

> The :class:`StackingClassifier` and :class:`StackingRegressor` provides such

Reviewer: provides -> provide

trained by cross-validation using internally
:func:`sklearn.model_selection.cross_val_predict`.

Note that the output of the ``estimators`` is controlled by the parameter
``method_estimators``. It corresponds to the method called by each
estimator. This parameter is either a list of string of the methods name or
``'auto'`` which will automatically called the method depending of the
availability and a pre-determined order of preference.

It is possible to get the final predictions using the ``predict`` or
``predict_proba`` depending if you are using :class:`StackingClassifier` or
:class:`StackingRegressor`::

   >>> y_pred = reg.predict(X_test)
   >>> from sklearn.metrics import r2_score
   >>> print('R2 score: {:.2f}'.format(r2_score(y_test, y_pred)))
   R2 score: 0.85

Note that it is also possible to get the output of the stacked outputs of the
``estimators`` using the ``transform`` method::

  >>> reg.transform(X_test[:5])  # doctest: +ELLIPSIS
  array([[28.78..., 28.43...  , 22.62...],
         [35.96..., 32.58..., 23.68...],
         [14.97..., 14.05..., 16.45...],
         [25.19..., 25.54..., 22.92...],
         [18.93..., 19.26..., 17.03... ]])

> 

Reviewer: maybe add a note that multiple layers can be achieved by setting `final_estimator=Stacking...(...)`

Optionally, weights can be provided for the individual classifiers::

   >>> eclf = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)], voting='soft', weights=[2,5,1])

.. _stacking:

Stacked generalization
----------------------

Stacked generalization is a method for combining estimators to reduce their

> biases [W1992]. More precisely, the predictions of each individual estimators

Reviewer 4: estimator

are stacked together and used by another estimator to compute the final
prediction. This final estimator is trained through cross-validation.

The :class:`StackingClassifier` and :class:`StackingRegressor` provides such

> strategies which can be applied to classification and regression problem.

Reviewer 4: problems

The ``estimators`` parameter corresponds to the list of the estimators which

> are stacked together. It should be given as a list of name and instance of

Reviewer 4: names and estimators?

estimators::

  >>> from sklearn.linear_model import RidgeCV, LassoCV
  >>> from sklearn.svm import SVR
  >>> estimators = [('ridge', RidgeCV()),
  ...               ('lasso', LassoCV(random_state=42)),
  ...               ('svr', SVR(C=1, gamma=1e-6, kernel='rbf'))]

> The ``final_estimator`` will combine the prediction of the ``estimators``. It

Reviewer 4: predictions

needs to be a classifier or a regressor when using :class:`StackingClassifier`
or :class:`StackingRegressor`, respectively::

  >>> from sklearn.ensemble import GradientBoostingRegressor
  >>> from sklearn.ensemble import StackingRegressor
  >>> reg = StackingRegressor(
  ...     estimators=estimators,
  ...     final_estimator=GradientBoostingRegressor(random_state=42))

To train the ``estimators`` and ``final_estimator``, the ``fit`` method needs
to be called on the training data::

  >>> from sklearn.datasets import load_boston
  >>> X, y = load_boston(return_X_y=True)
  >>> from sklearn.model_selection import train_test_split
  >>> X_train, X_test, y_train, y_test = train_test_split(X, y,
  ...                                                     random_state=42)
  >>> reg.fit(X_train, y_train)  # doctest: +ELLIPSIS
  StackingRegressor(...)

> During this fitting, the ``estimators`` are fitted on the whole training data

Reviewer 4: During the training?

Optionally, weights can be provided for the individual classifiers::

>    >>> eclf = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2), ('gnb', clf3)], voting='soft', weights=[2,5,1])

Reviewer 5: flake8 error: E231 missing whitespace after ','

.. topic:: Examples:

  * :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor.py`

.. _stacking:

Stacked generalization
======================

Stacked generalization is a method for combining estimators to reduce their
biases [W1992]. More precisely, the predictions of each individual estimator
are stacked together and used as input to a final estimator to compute the
final prediction. This final estimator is trained through cross-validation.

The :class:`StackingClassifier` and :class:`StackingRegressor` provide such
strategies which can be applied to classification and regression problems.

> The `estimators` parameter corresponds to the list of the estimators which

Author: > Is it common to stack multiple models?

I think so. Which other use-case do you have in mind?

> The `estimators` parameter corresponds to the list of the estimators which

Reviewer 6: I thought the default was stacking one base model and one final model.

> The `estimators` parameter corresponds to the list of the estimators which

Author: I think that this is more common to stack multiple models on the first layer and use a final meta-learner

or :class:`StackingRegressor`, respectively::

  >>> from sklearn.ensemble import GradientBoostingRegressor
  >>> from sklearn.ensemble import StackingRegressor
  >>> reg = StackingRegressor(
  ...     estimators=estimators,
  ...     final_estimator=GradientBoostingRegressor(random_state=42))

To train the `estimators` and `final_estimator`, the `fit` method needs
to be called on the training data::

  >>> from sklearn.datasets import load_boston
  >>> X, y = load_boston(return_X_y=True)
  >>> from sklearn.model_selection import train_test_split
  >>> X_train, X_test, y_train, y_test = train_test_split(X, y,
  ...                                                     random_state=42)
  >>> reg.fit(X_train, y_train)
  StackingRegressor(...)

During training, the `estimators` are fitted on the whole training data
`X_train`. They will be used when calling `predict`, `predict`, or
`predict_proba`. To generalize and avoid over-fitting, the
`final_estimator` is trained on out-samples using internally
:func:`sklearn.model_selection.cross_val_predict`.

Note that the output of the `estimators` is controlled by the parameter
`predict_method`. It corresponds to the method called by each

> estimator. This parameter is either a list of strings, being method names, or

Reviewer 6: After reading the code and warning below, I now understand what this is doing and this description seems very unclear. also it seems pretty complicated. Is this a common use-case? Were the different usage scenarios for this discussed (sorry this is a long PR ;). I feel like giving a single string and "auto" should cover most use-cases and be easier to document and understand.
I could imagine there's someone that prefers ``decision_function`` over ``predict_proba`` and then maybe they want to change the order of priority, but I don't see why you would specify this per estimator.

> estimator. This parameter is either a list of strings, being method names, or

Author: Is there any classifier in practice that we don't want to use absolutely `predict` and `predict_proba` in which it would be an issue? Otherwise, I agree with your statement. We could add this case if really this is missing. Always easier in this order.

> estimator. This parameter is either a list of strings, being method names, or

Reviewer 6: My point is more to remove the current per-estimator support which I find verbose and not well motivated ;)

`'auto'` which will automatically identify an available method depending on the
availability and a pre-determined order of preference (`predict_proba`,
`decision_function` and `predict`).

The fitted stacking estimators have the `predict` method, and classifier also
have the `predict_proba` method. They can be used as any other
estimator, e.g.::

   >>> y_pred = reg.predict(X_test)
   >>> from sklearn.metrics import r2_score
   >>> print('R2 score: {:.2f}'.format(r2_score(y_test, y_pred)))
   R2 score: 0.81

Note that it is also possible to get the output of the stacked outputs of the
`estimators` using the `transform` method::

  >>> reg.transform(X_test[:5])
  array([[28.78..., 28.43...  , 22.62...],
         [35.96..., 32.58..., 23.68...],
         [14.97..., 14.05..., 16.45...],
         [25.19..., 25.54..., 22.92...],
         [18.93..., 19.26..., 17.03... ]])

.. note::
   Multiple stacking layers can be achieved by assigning `final_estimator` to
   a :class:`StackingClassifier` or :class:`StackingRegressor`.

.. topic:: References

>    .. [W1992] Wolpert, David H. "Stacked generalization." Neural networks 5.2

Author: I'll check tomorrow but I am unsure if there is a stacking part in ESL.

>    .. [W1992] Wolpert, David H. "Stacked generalization." Neural networks 5.2

Reviewer 7: 8.8 Model Averaging and Stacking (begins on pg. 288 in ESL). So pumped for this!

>    .. [W1992] Wolpert, David H. "Stacked generalization." Neural networks 5.2

Reviewer 6: might be for a single model only, though? But definitely more readable than wolpert.

>    .. [W1992] Wolpert, David H. "Stacked generalization." Neural networks 5.2

Author: @jmwoloso Thanks, this is going faster like this :)

>    .. [W1992] Wolpert, David H. "Stacked generalization." Neural networks 5.2

Reviewer 7: @glemaitre no problem! really looking forward to this. let me know how else I can assist.

>    .. [W1992] Wolpert, David H. "Stacked generalization." Neural networks 5.2

Reviewer 7: i've also run across the following sources where stacking is discussed, but haven't read them myself, but rather came across them in the vtreat paper (quick read and really awesome stuff as well; they also just released [pyvtreat](https://github.com/WinVector/pyvtreat)). links below if needed/useful. seems i've read before over the years that there are multiple ways to do stacking. it was always on my to-do list to become more familiar with the details. this will be a good excuse :D

[vtreat paper](https://arxiv.org/pdf/1611.09477.pdf)

[super learner in prediction](https://biostats.bepress.com/ucbbiostat/paper266/)

[on cross validation and stacking](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.188.2952&rep=rep1&type=pdf)

[general oracle inequalities for model selection](https://projecteuclid.org/euclid.ejs/1236089916)

.. topic:: Examples:

  * :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor.py`

.. _stacking:

Stacked generalization
======================

Stacked generalization is a method for combining estimators to reduce their
biases [W1992]_ [HTF]_. More precisely, the predictions of each individual
estimator are stacked together and used as input to a final estimator to
compute the prediction. This final estimator is trained through
cross-validation.

The :class:`StackingClassifier` and :class:`StackingRegressor` provide such
strategies which can be applied to classification and regression problems.

The `estimators` parameter corresponds to the list of the estimators which
are stacked together in parallel on the input data. It should be given as a
list of names and estimators::

  >>> from sklearn.linear_model import RidgeCV, LassoCV
  >>> from sklearn.svm import SVR
  >>> estimators = [('ridge', RidgeCV()),
  ...               ('lasso', LassoCV(random_state=42)),

>   ...               ('svr', SVR(C=1, gamma=1e-6, kernel='rbf'))]

Reviewer 5: kernel='rbf' is redundant.
